### PR TITLE
log trigger based upkeep for basic test

### PR DESCRIPTION
**Pull request by shileiwill**
This PR should be on top of the ocr3 config changes [PR](https://github.com/smartcontractkit/chainlink/pull/10169).

The goal of this PR is to enable log based upkeeps in integration test. Per chatting with Anirudh, we are able to see `Added logs to buffer` in logs, so logs were indeed emitted and seen, but Perform didn't happen.

Open this PR for comments.

-------
Update: PR is ready for review. v21 tests are not enabled for now, will enable when e2e, with offchain code, is ready.

### CODE REVIEW

File: integration-tests/smoke/automation_test.go [hunk 1]
```diff
@@ -143,13 +148,20 @@ func SetupAutomationBasic(t *testing.T, nodeUpgrade bool) {
...
+				isLogTrigger,
 			)
 
 			l.Info().Msg("Waiting for all upkeeps to be performed")
 			gom := gomega.NewGomegaWithT(t)
 			gom.Eventually(func(g gomega.Gomega) {
 				// Check if the upkeeps are performing multiple times by analyzing their counters and checking they are greater than 10
 				for i := 0; i < len(upkeepIDs); i++ {
+					// Kick off the upkeep, and count!
+					err := consumers[i].Start()
```
Comment (shileiwill) [inline integration-tests/smoke/automation_test.go:160]: This is how we kick off the log trigger and count.

**Review by anirudhwarrier — CHANGES_REQUESTED** [review]

File: integration-tests/contracts/contract_deployer.go [hunk 2]
```diff
@@ -1093,6 +1103,25 @@ func (e *EthereumContractDeployer) DeployKeeperConsumer(updateInterval *big.Int)
 	}, err
 }
 
+func (e *EthereumContractDeployer) DeployAutomationrLogTriggerConsumer(updateInterval *big.Int) (KeeperConsumer, error) {
```
Comment (RyanRHall) [inline integration-tests/contracts/contract_deployer.go:1106]: typo `DeployAutomationLogTriggerConsumer`

File: integration-tests/contracts/ethereum_keeper_contracts.go [hunk 3]
```diff
@@ -1724,25 +1759,61 @@ func (v *EthereumKeeperRegistrar) EncodeRegisterRequest(
...
-			common.HexToAddress(adminAddr),
-			uint8(0), // trigger type
-			checkData,
-			[]byte{}, // triggerConfig
-			[]byte{}, // offchainConfig
-			amount,
-			common.HexToAddress(senderAddr),
-		)
+		if isLogTrigger {
+			logTriggerConfigType := goabi.MustNewType("tuple(address contractAddress, uint8 filterSelector, bytes32 topic0, bytes32 topic1, bytes32 topic2, bytes32 topic3)")
```
Comment (RyanRHall) [inline integration-tests/contracts/ethereum_keeper_contracts.go:1763]: can we use the automationUtils contract to pack this struct instead? It adds much better type safety :)

File: integration-tests/smoke/automation_test.go [hunk 4]
```diff
@@ -103,8 +103,9 @@ func SetupAutomationBasic(t *testing.T, nodeUpgrade bool) {
...
 
 	registryVersions := map[string]ethereum.KeeperRegistryVersion{
-		"registry_2_0": ethereum.RegistryVersion_2_0,
-		//"registry_2_1": ethereum.RegistryVersion_2_1,
+		"registry_2_0":             ethereum.RegistryVersion_2_0,
+		"registry_2_1_conditional": ethereum.RegistryVersion_2_1,
+		"registry_2_1_logtrigger":  ethereum.RegistryVersion_2_1,
 	}
 
 	for n, rv := range registryVersions {
```
Comment (RyanRHall) [inline integration-tests/smoke/automation_test.go:111]: why not just do...
```go
for name, registryVersion := range registryVersions {
```

File: integration-tests/smoke/automation_test.go (hunk 1 above)
Comment (shileiwill) [reply integration-tests/smoke/automation_test.go:160]: Both conditional and log based upkeeps can run e2e pretty stably. I triggered this CI test 5 times, and the `ETH smoke tests automation` passed all of them.

File: integration-tests/smoke/automation_test.go [hunk 5]
```diff
@@ -133,6 +134,10 @@ func SetupAutomationBasic(t *testing.T, nodeUpgrade bool) {
 				return
 			}
 
+			// Use the name to determine if this is a log trigger or not
+			// Not an elegant solution but it reuses the most code and avoids a lot of duplication
+			isLogTrigger := name == "registry_2_1_logtrigger"
```
Comment (RyanRHall) [inline integration-tests/smoke/automation_test.go:139]: seems fine to me! I would maybe get rid of the comment though. Feels like it doesn't add much.

**Review by RyanRHall — CHANGES_REQUESTED** [review]
looks good! just a couple small comments

File: integration-tests/smoke/automation_test.go (hunk 5 above)
Comment (shileiwill) [reply integration-tests/smoke/automation_test.go:139]: haha, it is so hacky that i want to give my reasoning :) I will simplify it.

File: integration-tests/contracts/ethereum_keeper_contracts.go [hunk 6]
```diff
@@ -1724,25 +1759,113 @@ func (v *EthereumKeeperRegistrar) EncodeRegisterRequest(
...
+				amount,
+				common.HexToAddress(senderAddr),
+			)
+
+			if err != nil {
+				return nil, err
+			}
+			return req, nil
+
+			//logTriggerConfigType := goabi.MustNewType("tuple(address contractAddress, uint8 filterSelector, bytes32 topic0, bytes32 topic1, bytes32 topic2, bytes32 topic3)")
```
Comment (shileiwill) [inline integration-tests/contracts/ethereum_keeper_contracts.go:1816]: I kept this for comparison, PR is not ready for landing.

@RyanRHall  , I am trying to use the utils but with the change, upkeeps dont perform, I believe its bc the topics dont match. I think the byte[32] aligns with the original hex string. Could you help double check?

File: integration-tests/contracts/ethereum_keeper_contracts.go [hunk 7]
```diff
@@ -1724,25 +1759,113 @@ func (v *EthereumKeeperRegistrar) EncodeRegisterRequest(
...
+
+			logTriggerConfigStruct := automation_utils_2_1.LogTriggerConfig{
+				ContractAddress: common.HexToAddress(upkeepAddr),
+				FilterSelector:  0,
+				Topic0:          topic0InBytes,
+				Topic1:          topic1InBytes,
+				Topic2:          topic2InBytes,
+				Topic3:          topic3InBytes,
+			}
+			encodedLogTriggerConfig, err := utilsABI.Pack("_logTriggerConfig", &logTriggerConfigStruct)
```
Comment (RyanRHall) [inline integration-tests/contracts/ethereum_keeper_contracts.go:1791]: you have to remove the first 4 bytes :( bad develop experience I know.... :/ maybe there's a better way to do it?

**Review by anirudhwarrier — APPROVED** [review]

File: integration-tests/contracts/contract_deployer.go [hunk 8]
```diff
@@ -1150,6 +1160,25 @@ func (e *EthereumContractDeployer) DeployKeeperConsumer(updateInterval *big.Int)
 	}, err
 }
 
+func (e *EthereumContractDeployer) DeployAutomationLogTriggerConsumer(updateInterval *big.Int) (KeeperConsumer, error) {
```
Comment (anirudhwarrier) [inline integration-tests/contracts/contract_deployer.go:1163]: `updateInterval` should be `testInterval`

File: integration-tests/actions/keeper_helpers.go [hunk 9]
```diff
@@ -419,27 +393,30 @@ func RegisterUpkeepContractsWithCheckData(
...
 
 	for contractCount := 0; contractCount < numberOfContracts; contractCount++ {
 		// Deploy consumer
-		keeperConsumerInstance, err := contractDeployer.DeployKeeperConsumer(big.NewInt(5))
-		require.NoError(t, err, "Deploying KeeperConsumer instance %d shouldn't fail", contractCount+1)
+		var keeperConsumerInstance contracts.KeeperConsumer
+		var err error
+
+		if isLogTrigger {
+			keeperConsumerInstance, err = contractDeployer.DeployAutomationLogTriggerConsumer(big.NewInt(5))
```
Comment (anirudhwarrier) [inline integration-tests/actions/keeper_helpers.go:406]:
```suggestion
			keeperConsumerInstance, err = contractDeployer.DeployAutomationLogTriggerConsumer(big.NewInt(1000))   //1000 block test range
```

**Review by anirudhwarrier — CHANGES_REQUESTED** [review]
https://github.com/smartcontractkit/chainlink/pull/10159/files#r1300478757
https://github.com/smartcontractkit/chainlink/pull/10159/files#r1300476939

**Review by FelixFan1992 — APPROVED** [review]

File: integration-tests/contracts/ethereum_keeper_contracts.go [hunk 10]
```diff
@@ -239,10 +231,9 @@ func (rcs *KeeperRegistrySettings) EncodeOnChainConfig(registrar string) ([]byte
 			Registrars:             []common.Address{common.HexToAddress(registrar)},
 			UpkeepPrivilegeManager: common.Address{},
 		}
-		encodedOnchainConfig, err := utilsABI.Pack("_onChainConfig", &onchainConfigStruct)
+		encodedOnchainConfig, err := utilsABI.Methods["_onChainConfig"].Inputs.Pack(&onchainConfigStruct)
```
Comment (RyanRHall) [inline integration-tests/contracts/ethereum_keeper_contracts.go:234]: nice :sunglasses:

File: integration-tests/smoke/automation_test.go [hunk 11]
```diff
@@ -104,12 +104,11 @@ func SetupAutomationBasic(t *testing.T, nodeUpgrade bool) {
 
 	registryVersions := map[string]ethereum.KeeperRegistryVersion{
 		"registry_2_0": ethereum.RegistryVersion_2_0,
-		//"registry_2_1": ethereum.RegistryVersion_2_1,
+		//"registry_2_1_conditional": ethereum.RegistryVersion_2_1,
+		//"registry_2_1_logtrigger": ethereum.RegistryVersion_2_1,
```
Comment (RyanRHall) [inline integration-tests/smoke/automation_test.go:108]: can we add a comment here with a TODO that points to the ticket to enable these tests? I just don't want us to forget to turn them on lol

**Review by RyanRHall — CHANGES_REQUESTED** [review]

File: integration-tests/contracts/ethereum_keeper_contracts.go (hunk 10 above)
Comment (ogtownsend) [reply integration-tests/contracts/ethereum_keeper_contracts.go:234]: Is `.Methods` generally considered safer than `.Pack`?

**Review by RyanRHall — APPROVED** [review]

File: integration-tests/contracts/ethereum_keeper_contracts.go (hunk 10 above)
Comment (shileiwill) [reply integration-tests/contracts/ethereum_keeper_contracts.go:234]: When we just use the Pack function, we have to slice off the first 4 bytes which is the function signature. We dont like the magic number 4 and it is easy to forget too.
Using this Methods function, we dont need to remember the magic 4 bytes any more :)

File: integration-tests/smoke/automation_test.go (hunk 11 above)
Comment (shileiwill) [reply integration-tests/smoke/automation_test.go:108]: https://smartcontract-it.atlassian.net/browse/AUTO-4617 will work on it tomorrow before Anirudh taking PTO :)

**Pull request merged**